File: data/code_pairs/cpp/2259.cpp
// descramble_matrix.cpp
#include <iostream>
#include <vector>

void descramble_matrix(std::vector<std::vector<double>>& mat, int na, const std::vector<int>& nfps, const std::vector<int>& ixshells, int nshells) {
    if (na <= 0) {
        return;
    }

    std::vector<int> nfps_orig(nshells, 0);
    std::vector<double> xa(na);
    std::vector<int> ixa(na);

    for (int i = 0; i < nshells; ++i) {
        nfps_orig[ixshells[i] - 1] = nfps[i];
    }

    int istart = 0;
    for (int i = 0; i < nshells; ++i) {
        int ishell = ixshells[i];

        int n = 0;
        for (int j = 0; j < ishell - 1; ++j) {
            n += nfps_orig[j];
        }

        for (int j = 0; j < nfps_orig[ishell - 1]; ++j) {
            ixa[istart] = n + j;
            ++istart;
        }
    }

    for (int j = 0; j < na; ++j) {
        for (int i = 0; i < na; ++i) {
            xa[i] = mat[i][j];
        }

        for (int i = 0; i < na; ++i) {
            mat[ixa[i]][j] = xa[i];
        }
    }

    for (int i = 0; i < na; ++i) {
        for (int j = 0; j < na; ++j) {
            xa[j] = mat[i][j];
        }

        for (int j = 0; j < na; ++j) {
            mat[i][ixa[j]] = xa[j];
        }
    }
}

int main() {
    int na = 4;
    int nshells = 2;
    std::vector<int> nfps = {2, 2};
    std::vector<int> ixshells = {1, 2};

    std::vector<std::vector<double>> mat = {
        {1.0, 2.0, 3.0, 4.0},
        {5.0, 6.0, 7.0, 8.0},
        {9.0, 10.0, 11.0, 12.0},
        {13.0, 14.0, 15.0, 16.0}
    };

    descramble_matrix(mat, na, nfps, ixshells, nshells);

    std::cout << "Matrix after descrambling:" << std::endl;
    for (const auto& row : mat) {
        for (const auto& val : row) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}

File: data/code_pairs/fortran/2259.f90
! descramble_matrix.f90
subroutine descramble_matrix(mat, na, nfps, ixshells, nshells)
    implicit none
    integer, intent(in) :: na, nshells
    integer, intent(in) :: nfps(nshells), ixshells(nshells)
    double precision, intent(inout) :: mat(na, na)
    double precision :: xa(na)
    integer :: ixa(na)
    integer :: nfps_orig(nshells)
    integer :: i, j, istart, ishell, n

    if (na .gt. 0) then
        do i = 1, nshells
            nfps_orig(ixshells(i)) = nfps(i)
        end do

        istart = 1
        do i = 1, nshells
            ishell = ixshells(i)

            n = 0
            do j = 1, ishell-1
                n = n + nfps_orig(j)
            end do

            do j = 1, nfps_orig(ishell)
                ixa(istart) = n + j
                istart = istart + 1
            end do
        end do

        do j = 1, na
            do i = 1, na
                xa(i) = mat(i, j)
            end do

            do i = 1, na
                mat(ixa(i), j) = xa(i)
            end do
        end do

        do i = 1, na
            do j = 1, na
                xa(j) = mat(i, j)
            end do

            do j = 1, na
                mat(i, ixa(j)) = xa(j)
            end do
        end do
    endif
end subroutine descramble_matrix